README.rst
service_person_stamdata_udvidet
*******************************
Module that integrates with the service_person_stamdata_udvidet service from Serviceplatformen.

:Author:
    Heini Leander Ovason <heini@magenta.dk>


Quick notes regarding use of the test and production service webapi
===================================================================
Please note that access to the service webapi depends on the level of service agreement in place. 

The service agreement may only grant access to either the test or the production service. 
(It may also be possible to have an agreement which grants access to both systems)

**Thus access to the production service does not automatically
constitue access to the test service.**

The api exposes a toggle to access either the production or the test system (production by default):

.. code-block:: python


    get_citizen(..., production=<True|False>)


Additionally it is also important to note that it is not possible
to use "real" cpr identifiers on the test system and vice versa. 



API (with examples)
===================

*In order for this module to interact with Serviceplatformen you need valid 'Invocation context' UUIDs and a certificate.*

get_citizen()
-------------
*The function serves as a facade to ease interaction with Serviceplatformen's 'SF1520 - Udvidet person stamdata (lokal)' service.*

.. code-block:: python

    # -- coding: utf-8 --
    import json

    from service_person_stamdata_udvidet import  get_citizen

    uuids = {
        'service_agreement': '42571b5d-6371-4edb-8729-1343a3f4c9b9',
        'user_system': '99478e20-68e6-41ff-b822-681fb69b8ff2',
        'user': 'e3108916-8ed9-4482-8045-7b46c83904b0',
        'service': '9883c483-d42f-424a-9a2a-94d1d200d294'
    }

    certificate = '/path/to/certificate.crt'

    cprnr = '0123456789'

    result = get_citizen(
        service_uuids=uuids,
        certificate=certificate,
        cprnr=cprnr,
        production=True  # Set this to `False` to access the test system
    )

    print(json.dumps(result))

`Example Output <https://pastebin.com/MSmk3YaB>`_

test can be run from this directory by issuing 

    python -m unittest discover




CHANGELOG.rst
=========
Changelog
=========

Module that integrates with the service_person_stamdata_udvidet service from Serviceplatformen.
Please review the official docs for the service:

  https://www.serviceplatformen.dk/administration/serviceOverview/show?uuid=e6be2436-bf35-4df2-83fe-925142825dc2


Version 0.2.1 (February 21 2021)
---------------------

* Fix deprecation warning in docstring

Version 0.2.0 (July 15 2019)
---------------------

* 30041: Updated service endpoints to comply with api version 4.0


Version 0.1.3 (Oct 30, 2018)
---------------------

* don't install the tests


Version 0.1.1 (Jun 21, 2018)
---------------------

* relax dependencies to only what's strictly required


Version 0.1.0 (May 2, 2018)
---------------------

* Release version 0.1.0


service_person_stamdata_udvidet/helpers.py
# -- coding: utf-8 --
#
# Copyright (c) 2017, Magenta ApS
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import re
import requests
from jinja2 import Template

__author__ = "Heini Leander Ovason"


def construct_envelope_SF1520(template, service_uuids, cprnr):
    """The function returnes a envelope for the service
    'SF1520 - Udvidet person stamdata (lokal)'."""

    with open(template, "r") as filestream:
        template_string = filestream.read()

    xml_template = Template(template_string)

    populated_template = xml_template.render(
        cprnr=cprnr,
        service_agreement=service_uuids['service_agreement'],
        user_system=service_uuids['user_system'],
        user=service_uuids['user'],
        service=service_uuids['service']
    )

    # service platform requirement.
    latin_1_encoded_soap_envelope = populated_template.encode('latin-1')

    return latin_1_encoded_soap_envelope


def construct_envelope_SF6002(
    template,
    service_uuids,
    cprnr,
    operation,
    parameter_type
):
    """The function returnes a envelope for the service
    'SF6002 - CPR Abonnement'."""

    with open(template, "r") as filestream:
        template_string = filestream.read()

    xml_template = Template(template_string)

    populated_template = xml_template.render(
        cprnr=cprnr,
        service_agreement=service_uuids['service_agreement'],
        user_system=service_uuids['user_system'],
        user=service_uuids['user'],
        service=service_uuids['service'],
        operation=operation,
        parameter_type=parameter_type
    )

    # service platform requirement.
    latin_1_encoded_soap_envelope = populated_template.encode('latin-1')

    return latin_1_encoded_soap_envelope


def http_post(endpoint, soap_envelope, certificate):

    if not endpoint and not soap_envelope and not certificate:
        return None

    response = requests.post(
        url=endpoint,
        data=soap_envelope,
        cert=certificate
    )

    return response


def validate_cprnr(cprnr):

    if not cprnr:
        # Log e.g. 'Type error occured: input'
        return False

    check = re.match(r'^\d{10}$', cprnr)

    if not check:
        # Log e.g. 'Not a valid cprnr'
        return False

    return True



service_person_stamdata_udvidet/udvidet_person_stam_data_lokal.py
# -- coding: utf-8 --
#
# Copyright (c) 2017, Magenta ApS
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import os
import xmltodict
from .helpers import http_post
from .helpers import validate_cprnr
from .helpers import construct_envelope_SF1520

__author__ = "Heini Leander Ovason"


# Service
prod_service_url = (
    "https://prod.serviceplatformen.dk/service/CPR/PersonBaseDataExtended/4"
)
test_service_url = (
    "https://exttest.serviceplatformen.dk/service/CPR/PersonBaseDataExtended/4"
)


def get_citizen(service_uuids, certificate, cprnr, production=True):
    r"""
    The function returnes a citizen dict from the
    'SF1520 - Udvidet person stamdata (lokal)' service.
    It serves as a facade to simplify input validation, and interaction
    with the SOAP service, parsing and filtering the response.

    :param dict service_uuids: The required service uuids

    .. note::

        Here is sample of the service_uuid dictionary:
        
            {
                'service_agreement': '42571b5d-6371-4edb-8729-1343a3f4c9b9',
                'user_system': '99478e20-68e6-41ff-b822-681fb69b8ff2',
                'user': 'e3108916-8ed9-4482-8045-7b46c83904b0',
                'service': '9883c483-d42f-424a-9a2a-94d1d200d294'
            }

    :param str certificate: Path to the required certificate bundle

    :param str cprnr:  String of 10 digits -> r'^\d{10}$'
    
    :param bool production: Production/Test service mode toggle 

    .. note::

        Production and Test services are seperate systems hosted by
        serviceplatformen.

        Please note that it is NOT possible to use "real" cpr identifiers
        on the test platform and vice versa.


    :return: Dictionary representation of a citizen
    :rtype: dict
    """

    if production:
        service_url = prod_service_url
    else:
        service_url = test_service_url

    is_cprnr_valid = validate_cprnr(cprnr)

    if is_cprnr_valid:

        template_directory = os.path.dirname(__file__)

        soap_envelope_template = os.path.join(
            template_directory, "soap_envelope_template.xml"
        )

        soap_envelope = construct_envelope_SF1520(
            template=soap_envelope_template, service_uuids=service_uuids, cprnr=cprnr
        )
        response = call_cpr_person_lookup_request(
            soap_envelope=soap_envelope,
            certificate=certificate,
            service_url=service_url,
        )
        if response.status_code == 200:
            citizen_dict = parse_cpr_person_lookup_xml_to_dict(
                soap_response_xml=response.text
            )
            return citizen_dict
        else:
            response.raise_for_status()
            return {"Error": "Something went wrong"}


def call_cpr_person_lookup_request(soap_envelope, certificate, service_url):
    """Performs a web service call to 'Udvidet Person Stam Data(lokal)'.
    : param soap_envelope: SOAP envelope
    : param certificate: Path to certificate
    : param service_url: url endpoint for service
    : type soap_envelope: str
    : type certificate: str
    : type service_url: str
    :return: Complete serviceplatform xml representation of a citizen
    :rtype: str"""

    response = http_post(
        endpoint=service_url, soap_envelope=soap_envelope, certificate=certificate
    )

    return response


def parse_cpr_person_lookup_xml_to_dict(soap_response_xml):
    """Parses string xml to a dict
    : param soap_response_xml: xml
    : type soap_response_xml: str
    : return: Dictionary representation of a citizen
    : rtype: dict"""

    namespaces = {
        "http://serviceplatformen.dk/xml/schemas/cpr/PNR/1/": None,
        "http://serviceplatformen.dk/xml/schemas/InvocationContext/1/": None,
        "http://serviceplatformen.dk/xml/schemas/AuthorityContext/1/": None,
        "http://serviceplatformen.dk/xml/schemas/CallContext/1/": None,
        "http://serviceplatformen.dk/xml/wsdl/soap11/CPR/PersonBaseDataExtended/4/": None,
        "http://schemas.xmlsoap.org/soap/envelope/": None,
    }
    xml_to_dict = xmltodict.parse(
        soap_response_xml, process_namespaces=True, namespaces=namespaces
    )
    root = xml_to_dict["Envelope"]["Body"]["PersonLookupResponse"]

    citizen_dict = {}

    name = root["persondata"]["navn"]
    for k, v in name.items():
        citizen_dict[k] = v

    person_data = root["persondata"]
    person_data.pop("navn")
    for k, v in person_data.items():
        citizen_dict[k] = v

    address = root["adresse"]["aktuelAdresse"]
    if not address:
        address = {}
    for k, v in address.items():
        citizen_dict[k] = v

    try:
        not_living_in_dk = root["adresse"]["udrejseoplysninger"]  # noqa F841
        citizen_dict["udrejst"] = True
    except KeyError:
        citizen_dict["udrejst"] = False

    relations = root["relationer"]
    citizen_dict["relationer"] = []
    for k, v in relations.items():
        # NOTE: v is a dict if k is:
        # 'mor', 'far', or 'aegtefaelle'.
        if isinstance(v, dict):
            citizen_dict["relationer"].append(
                {"relation": k, "cprnr": v.get("personnummer")}
            )
        # NOTE: v is a list of dicts if k is 'barn'.
        if isinstance(v, list):
            for child in v:
                citizen_dict["relationer"].append(
                    {"relation": k, "cprnr": child.get("personnummer")}
                )

    return citizen_dict
